cli/internal/state/state.go

```diff
@@ -186,6 +206,343 @@ func (s *State) Merge(other *State) (*State, error) {
 	return s, nil
 }
 
+// Validate validates the state against the given constraint set, which can be one of
+//   - PreCreate, which is the constraint set that should be enforced before "constellation create" is run.
+//   - PreInit, which is the constraint set that should be enforced before "constellation apply" is run.
+//   - PostInit, which is the constraint set that should be enforced after "constellation apply" is run.
+func (s *State) Validate(constraintSet StateConstraints) error {
+	v := validation.NewValidator()
+
+	switch constraintSet {
+	case PreCreate:
+		return v.Validate(s, validation.ValidateOptions{
+			OverrideConstraints: s.preCreateConstraints,
+		})
+	case PreInit:
+		return v.Validate(s, validation.ValidateOptions{
+			OverrideConstraints: s.preInitConstraints,
+		})
+	case PostInit:
+		return v.Validate(s, validation.ValidateOptions{
+			OverrideConstraints: s.postInitConstraints,
+		})
+	default:
+		return fmt.Errorf("unknown constraint set")
+	}
+}
+
+// preCreateConstraints are the constraints on the state that should be enforced
+// before a Constellation cluster is created.
+//
+// The constraints check if the state file version is valid,
+// and if all fields are empty, which is a requirement pre-create.
+func (s *State) preCreateConstraints() []*validation.Constraint {
+	return []*validation.Constraint{
+		// state version needs to be accepted by the parsing CLI.
+		validation.OneOf(s.Version, []string{Version1}).
+			WithFieldTrace(s, &s.Version),
+		// Infrastructure must be empty.
+		// As the infrastructure struct contains slices, we cannot use the
+		// Empty constraint on the entire struct. Instead, we need to check
+		// each field individually.
+		validation.Empty(s.Infrastructure.UID).
```

Comment:
nit / idea: looking at this it seems that the WithFieldTrace seems always wanted. Why not integrate it into the condition and change the syntax to:
```
validation.OneOf(s, &s.Version, []string{Version1})
```
it would have the number of code.

Reply:
Of course this would be preferred, but this doesn't work (without reflection in the constraint function) due to the pass-by-value semantics of Go. Theoretically, we could implement it with some hacks in the constraint functions that we would have to implicitly require in all constraints, which seemed like a bad practice to me. Especially for custom constraints. However, this is something I will keep an eye on in the future, I definitely agree with you in terms of ergonomics here.

Reply:
I'm not sure why this doesn't work. but I do see that it would require to nicely wrap the reflect logic for DRY:

```
// OneOf is a constraint that s is in the set of values p.
func OneOf[T comparable](doc any, s T, p []T) *Constraint {
	c := Constraint{
		Satisfied: func() *TreeError {
			for _, v := range p {
				if s == v {
					return nil
				}
			}
			return NewErrorTree(fmt.Errorf("%v must be one of %v", s, p))
		},
	}
	derefedField := pointerDeref(reflect.ValueOf(s))
	fieldRef := referenceableValue{
		value: derefedField,
		addr:  derefedField.UnsafeAddr(),
		_type: derefedField.Type(),
	}
	derefedDoc := pointerDeref(reflect.ValueOf(doc))
	docRef := referenceableValue{
		value: derefedDoc,
		addr:  derefedDoc.UnsafeAddr(),
		_type: derefedDoc.Type(),
	}
	return c.withTrace(docRef, fieldRef)
}
```

Reply:
Both `doc` and `s` get copied by value